### PR TITLE
Ensure that Task SDK supervisor closes all its subprocess handles correctly.

### PR DESCRIPTION
**Pull request by ashb**
If we keep any copies of the handles open then the selector loop in monitor
process will get stuck waiting on a read loop on a socket that never closes
(because its open in the same process still!).

Sadly I wasn't able to reproduce this behaviour in unit tests, but only when
running this code for real. (The main fix here is to pass `child_stderr` to
_close_unused_sockets too, the rest are drive-by tidy ups)

Since we never access `proc.stdout` or stderr on the class (they are closed
over in the callbacks only) I removed those properties as they aren't needed
and shouldn't be accessed directly as it would lead to garbled output.

Also add tests (and re-fix) the last-chance exception handling

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 1]
```diff
@@ -192,7 +203,7 @@ def exit(n: int) -> NoReturn:
             sys.stdout.flush()
         with suppress(ValueError, OSError):
             sys.stderr.flush()
-        with suppress(ValueError, OSError):
+        with suppress(ValueError, OSError, AttributeError):
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]: What could cause an `AttributeError` here?

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py [hunk 2]
```diff
@@ -131,17 +131,17 @@ def _reopen_std_io_handles(child_stdin, child_stdout, child_stderr):
...
+        ("stdout", child_stdout, "w"),
+        ("stderr", child_stderr, "w"),
     ):
         handle = getattr(sys, handle_name)
         try:
             fd = handle.fileno()
             os.dup2(sock.fileno(), fd)
-            if close:
-                handle.close()
+            # dup2 creates another open copy of the fh, we can close the "socket" copy of it.
```
Comment (kaxil) [inline task_sdk/src/airflow/sdk/execution_time/supervisor.py]:
```suggestion
            # dup2 creates another open copy of the fd, we can close the "socket" copy of it.
```

**Review by kaxil — APPROVED** [review]

File: task_sdk/src/airflow/sdk/execution_time/supervisor.py (hunk 1 above)
Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: pytest

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: Though that was before i switched it to the `capfd` fixture -- that might not be needed, I'll try without it and remove if it's not needed anymore

Comment (ashb) [reply task_sdk/src/airflow/sdk/execution_time/supervisor.py]: YEah not needed anymore. Will remove

**Review by amoghrajesh — COMMENTED** [review]
Nice!

**Pull request merged**